finetuner/tuner/dataset/samplers.py

```diff
@@ -63,12 +63,20 @@ def __init__(
         for key, val in self._class_to_labels.items():
             self._cls_group_counts.append([key, ceil(len(val) / num_items_per_class)])
 
-        # Approximate (max) length
-        self._length = ceil(
-            sum([x[1] for x in self._cls_group_counts]) / self._num_classes
-        )
+        # Prepare the current batches
+        self._prepare_batches()
 
     def __iter__(self) -> Iterator[Tuple[int, ...]]:
+        batches = self._batches
+
+        # Prepare next batches - the reason they are prepared in advance is so that
+        # len() gives the correct length
+        self._prepare_batches()
```

Comment:
I think the proper implementation is to have some lazy property for `length`. This is quite a smell